### PR TITLE
Convert HasExtension, HasCriticialExtension + some others to util::Status

### PR DESCRIPTION
**Pull request by Martin2112**
More refactoring to get rid of Cert::Status usages.

### CODE REVIEW

**Comment by Martin2112** [comment]
Rebased + conflicts fixed. PTAL.

**Comment by Martin2112** [comment]
Rebased + latest conflicts fixed.

**Comment by pphaneuf** [comment]
-- Gee, Brain, what do you want to do tonight?
-- The same thing we do every night, Pinky - rebase and fix conflicts!

**Comment by Martin2112** [comment]
I know and they're all my own conflicts.

**Comment by pphaneuf** [comment]
That's one reason why I sometimes send in only one change at a time, because I have the others in a branch that's *on top* of the one I sent for review (so it won't conflict, the only conflicts would be for changes when addressing comments, usually more minimal). But it's more git branch wrangling, and it loses a bit of "reviewing concurrency"...

**Comment by Martin2112** [comment]
Yes these are all pretty much separate, the conflicts were minimal.

**Comment by pphaneuf** [comment]
Of note: I started reviewing this PR a number of times now, but I've given up before getting to the end every time... It's a bit too big for my taste?

I might have liked better tackling `HasExtension` (in itself quite simple) and its repercussions (less simple), then another one that's in itself simple and its not-so-simple repercussions, etc.

**Comment by Martin2112** [comment]
I more or less did it by groups of logically related functionality. I think it t made it easier to work out that the logic hadn't changed, if not to review.

**Comment by pphaneuf** [comment]
Yeah, they're related, but for reviewing, they traverse more than one layer, so I need to check all the callers, *and* the callers callers, and at some point, I lose the plot. :stuck_out_tongue:

For example, that little change in behaviour in that other PR, where `status == TRUE` became something else that's *almost* the same, is much easier to fish out in a smaller sized diff. In a bigger diff like this one, I don't feel nearly as confident (which might very well be my own limitation!).

**Comment by pphaneuf** [comment]
Heck, I find #910 complicated and scary! :wink: 

But I'll buckle down and do both, stand by... :smiley:

**Comment by Martin2112** [comment]
I know what you mean but pulling apart this type of logic often doesn't work easily on any obvious axis. It's nearly done now at least.

File: cpp/client/ct.cc [hunk 1]
```diff
@@ -202,14 +202,14 @@ static bool PrecertChainToEntry(const cert_trans::PreCertChain& chain,
     return false;
   }
 
-  Cert::Status status =
+  StatusOr<bool> status =
```
Comment (AlCutter) [inline cpp/client/ct.cc]: Couple of while-you're-here ones:

Constable, constable?
/status/has_poison/ ?

File: cpp/client/ct.cc [hunk 2]
```diff
@@ -813,9 +813,11 @@ void WrapEmbedded() {
   CertChain chain(pem_chain);
   CHECK(chain.IsLoaded()) << cert_file
                           << " is not a valid PEM-encoded certificate chain";
-  CHECK_EQ(Cert::TRUE,
-           chain.LeafCert()->HasExtension(
-               cert_trans::NID_ctEmbeddedSignedCertificateTimestampList));
+  CHECK_EQ(true,
```
Comment (AlCutter) [inline cpp/client/ct.cc]: CHECK(chain...)

File: cpp/client/ssl_client.cc [hunk 3]
```diff
@@ -184,9 +184,17 @@ int SSLClient::VerifyCallback(X509_STORE_CTX* ctx, void* arg) {
 
   string serialized_scts;
   // First, see if the cert has an embedded proof.
-  if (chain.LeafCert()->HasExtension(
-          cert_trans::NID_ctEmbeddedSignedCertificateTimestampList) ==
-      Cert::TRUE) {
+  const StatusOr<bool> has_timestamp_list = chain.LeafCert()->HasExtension(
```
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: `has_embedded_proof` to line up with the comment?

File: cpp/client/ssl_client.cc [hunk 4]
```diff
@@ -201,10 +209,8 @@ int SSLClient::VerifyCallback(X509_STORE_CTX* ctx, void* arg) {
     }
     // Else look for the proof in a superfluous cert.
     // Let's assume the superfluous cert is always last in the chain.
```
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Probably need to move this comment with the assumption up to ~L190

File: cpp/log/cert.cc [hunk 5]
```diff
@@ -683,9 +687,10 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
   // Let's first check if the extension is present. This allows us to
   // distinguish between "NID not recognized" and the more harmless
   // "extension not found, found more than once or corrupt".
-  Cert::Status status = HasExtension(extension_nid);
-  if (status != TRUE)
-    return status;
+  StatusOr<bool> status = HasExtension(extension_nid);
```
Comment (AlCutter) [inline cpp/log/cert.cc]: const

File: cpp/log/cert.cc [hunk 6]
```diff
@@ -966,15 +971,20 @@ util::Status Cert::IsValidNameConstrainedIntermediateCa() const {
   // If it's not a CA cert or there is no name constraint extension then we
   // don't need to apply the rules any further
   StatusOr<bool> has_ca_constraint = HasBasicConstraintCATrue();
```
Comment (AlCutter) [inline cpp/log/cert.cc]: const here & next line

File: cpp/log/cert.cc [hunk 7]
```diff
@@ -966,15 +971,20 @@ util::Status Cert::IsValidNameConstrainedIntermediateCa() const {
...
+  StatusOr<bool> has_name_constraints = HasExtension(NID_name_constraints);
+
   if (!has_ca_constraint.ok() || !has_ca_constraint.ValueOrDie()
-      || HasExtension(NID_name_constraints) == Cert::FALSE) {
+      || !has_name_constraints.ok() || !has_name_constraints.ValueOrDie()) {
```
Comment (AlCutter) [inline cpp/log/cert.cc]: This is different - previously failing to get the `NID_name_constraints` extension would not have returned `OK`, but an `INVALID_ARGUMENT` below.

Arguably that's misleading, but perhaps we could do something like
```C++
if (!has_ca_constraint.ok()) {
  return has_ca_constraint.status();
}
if (!has_name_constraints.ok()) {
  return has_name_constraints.status();
}
if (!has_ca_constraint.ValueOrDie() || !has_name_constraints.ValueOrDie()) {
  return OK;
}
```

File: cpp/log/cert.cc [hunk 8]
```diff
@@ -966,15 +971,20 @@ util::Status Cert::IsValidNameConstrainedIntermediateCa() const {
...
     return util::Status::OK;
   }
 
   // So there now must be a CT extension and the name constraint must not be
   // in error
-  if (HasExtension(NID_name_constraints) != Cert::TRUE
-      || HasExtension(NID_ctNameConstraintNologIntermediateCa) != Cert::TRUE) {
+  StatusOr<bool> has_ct_nolog_intermediate =
```
Comment (AlCutter) [inline cpp/log/cert.cc]: const

File: cpp/log/cert.cc [hunk 9]
```diff
@@ -966,15 +971,20 @@ util::Status Cert::IsValidNameConstrainedIntermediateCa() const {
...
+      HasExtension(NID_ctNameConstraintNologIntermediateCa);
+
+  if (!has_name_constraints.ValueOrDie() || !has_ct_nolog_intermediate.ok() ||
```
Comment (AlCutter) [inline cpp/log/cert.cc]: I think that `has_name_constraints` must be `true` here, so maybe pull this out of the `if` and `CHECK(has_name_contraints.ValueOrDie());` ?

File: cpp/log/cert.cc [hunk 10]
```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
...
-    return Cert::ERROR;
+    return util::Status(Code::FAILED_PRECONDITION, "Cert not loaded");
   }
 
   const Cert* pre = PreCert();
 
   // (1) Check that the leaf contains the critical poison extension.
-  Cert::Status status = pre->HasCriticalExtension(cert_trans::NID_ctPoison);
-  if (status != Cert::TRUE)
+  StatusOr<bool> status = pre->HasCriticalExtension(cert_trans::NID_ctPoison);
```
Comment (AlCutter) [inline cpp/log/cert.cc]:
```C++
const StatusOr<bool> has_poison(...);
```

File: cpp/log/cert.cc [hunk 11]
```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
...
+  if (!status.ok() || !status.ValueOrDie()) {
     return status;
+  }
 
   // (2) If signed by a Precertificate Signing Certificate, check that
   // the AKID extensions are compatible.
-  status = UsesPrecertSigningCertificate();
-  if (status == Cert::FALSE) {
+  StatusOr<bool> uses_precert_signing = UsesPrecertSigningCertificate();
```
Comment (AlCutter) [inline cpp/log/cert.cc]: const

File: cpp/log/cert.cc [hunk 12]
```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
...
+  if (uses_precert_signing.ok() && !uses_precert_signing.ValueOrDie()) {
     // If there is no precert signing extendedKeyUsage, no more checks:
     // the cert was issued by a regular CA.
-    return Cert::TRUE;
+    return true;
   }
-  if (status != Cert::TRUE)
+  if (!uses_precert_signing.ok() || !uses_precert_signing.ValueOrDie()) {
```
Comment (AlCutter) [inline cpp/log/cert.cc]:
```C++
if (!uses_precert_signing.ok()) {
  return uses_precert_signing.status();  // incorrectly returning status currently!
}

CHECK(uses_precert_signing.ValueOrDie());
```
?

File: cpp/log/cert.cc [hunk 13]
```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
...
+  }
 
   const Cert* issuer = PrecertIssuingCert();
   // If pre has the extension set but the issuer doesn't, error.
-  status = pre->HasExtension(NID_authority_key_identifier);
-  if (status == Cert::FALSE)
-    return Cert::TRUE;
-  if (status != Cert::TRUE)
-    return status;
+  StatusOr<bool> has_akid = pre->HasExtension(NID_authority_key_identifier);
```
Comment (AlCutter) [inline cpp/log/cert.cc]: const

File: cpp/log/cert.cc [hunk 14]
```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
...
+
+  if (has_akid.ok() && !has_akid.ValueOrDie()) {
+    return true;
+  }
+  if (!has_akid.ok()) {
+    return has_akid;
+  }
```
Comment (AlCutter) [inline cpp/log/cert.cc]: Could add `CHECK(has_akid.ValueOrDie())` here to guard against future modifications to this getting it wrong/document the invariant.

File: cpp/log/cert_checker.cc [hunk 15]
```diff
@@ -141,14 +141,15 @@ Status CertChecker::CheckCertChain(CertChain* chain) const {
     return Status(util::error::INVALID_ARGUMENT, "invalid certificate chain");
 
   // Weed out things that should obviously be precert chains instead.
-  Cert::Status status =
+  StatusOr<bool> status =
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: const
/status/has_poison/

File: cpp/log/cert_checker.cc [hunk 16]
```diff
@@ -178,12 +179,15 @@ Status CertChecker::CheckIssuerChain(CertChain* chain) const {
...
                                       string* issuer_key_hash,
                                       string* tbs_certificate) const {
-  if (!chain || !chain->IsLoaded())
+  if (!chain || !chain->IsLoaded()) {
     return Status(util::error::INVALID_ARGUMENT, "invalid certificate chain");
-  Cert::Status status = chain->IsWellFormed();
-  if (status == Cert::FALSE)
+  }
+
+  StatusOr<bool> status = chain->IsWellFormed();
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]:
```C++
const StatusOr<bool> chain_well_formed(chain->…);
```

File: cpp/log/cert_checker.cc [hunk 17]
```diff
@@ -178,12 +179,15 @@ Status CertChecker::CheckIssuerChain(CertChain* chain) const {
...
+  if (status.ok() && !status.ValueOrDie()) {
     return Status(util::error::INVALID_ARGUMENT, "prechain not well formed");
-  if (status != Cert::TRUE) {
+  }
+  if (!status.ok() || !status.ValueOrDie()) {
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]:
```C++
if (!chain_well_formed.ok()) {
 …
}

CHECK(chain_well_formed.ValueOrDie());
```

File: cpp/log/cert_checker.cc [hunk 18]
```diff
@@ -208,12 +212,13 @@ Status CertChecker::CheckPreCertChain(PreCertChain* chain,
       return status;
   }
 
-  Cert::Status uses_pre_issuer = chain->UsesPrecertSigningCertificate();
-  if (uses_pre_issuer != Cert::TRUE && uses_pre_issuer != Cert::FALSE)
+  StatusOr<bool> uses_pre_issuer = chain->UsesPrecertSigningCertificate();
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: const

File: cpp/log/cert_submission_handler.cc [hunk 19]
```diff
@@ -32,14 +33,14 @@ bool CertSubmissionHandler::X509ChainToEntry(const CertChain& chain,
   if (!chain.IsLoaded())
     return false;
 
-  Cert::Status status = chain.LeafCert()->HasExtension(
+  StatusOr<bool> status = chain.LeafCert()->HasExtension(
```
Comment (AlCutter) [inline cpp/log/cert_submission_handler.cc]: const ... has_embedded_proof

File: cpp/log/cert_submission_handler.cc [hunk 20]
```diff
@@ -133,9 +134,9 @@ bool CertSubmissionHandler::SerializedTbs(const Cert& cert, string* result) {
     return false;
   }
 
-  Cert::Status status = cert.HasExtension(
+  StatusOr<bool> status = cert.HasExtension(
```
Comment (AlCutter) [inline cpp/log/cert_submission_handler.cc]: `const … has_embedded_proof`

**Comment by AlCutter** [comment]
All done, that wasn't so bad :)

File: cpp/log/cert.cc (hunk 7 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: I think you're right. While it's slightly ambiguous I don't think it makes sense to return OK if the extension couldn't be looked up as it's only going to fail later when we read the data for it. That shouldn't happen anyway as it was registered at startup. We already checked the cert was loaded, which is the only other error case.

File: cpp/log/cert_checker.cc (hunk 17 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Not sure about that one. ValueOrDie() only dies if status is not OK. We want to return an error status if the chain isn't well formed.

**Comment by Martin2112** [comment]
That hopefully addresses all the comments.

File: cpp/client/ct.cc (hunk 1 above)
Comment (Martin2112) [inline cpp/client/ct.cc]: Done.

File: cpp/client/ct.cc (hunk 2 above)
Comment (Martin2112) [inline cpp/client/ct.cc]: Done.

File: cpp/client/ssl_client.cc (hunk 3 above)
Comment (Martin2112) [inline cpp/client/ssl_client.cc]: Done.

File: cpp/client/ssl_client.cc (hunk 4 above)
Comment (Martin2112) [inline cpp/client/ssl_client.cc]: Done. Split it into two parts, one where we get the value and one where the decision occurs.

File: cpp/log/cert.cc (hunk 5 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 6 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 7 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 8 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 9 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 10 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 11 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 12 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Fixed up the incorrect return. Didn't add a CHECK as this shouldn't exit if it's not well formed.

File: cpp/log/cert.cc (hunk 13 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert.cc (hunk 14 above)
Comment (Martin2112) [inline cpp/log/cert.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 15 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 16 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 17 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Didn't do this. It returns a non OK status unless the chain is well formed. I could add the CHECK as an invariant guarantee but I think the if statement is correct.

File: cpp/log/cert_checker.cc (hunk 18 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_submission_handler.cc (hunk 19 above)
Comment (Martin2112) [inline cpp/log/cert_submission_handler.cc]: Done.

File: cpp/log/cert_submission_handler.cc (hunk 20 above)
Comment (Martin2112) [inline cpp/log/cert_submission_handler.cc]: Done.

**Comment by Martin2112** [comment]
Replied to all comments.

File: cpp/log/cert_checker.cc (hunk 17 above)
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: I think it's right; if `chain_well_formed` was false (and ok), then we'd have returned at that point, so at 190 we only have to check if there was an error with the `!ok()`.

Comment (Martin2112) [inline cpp/log/cert_checker.cc]: I think you're right. Fixing it now and running the tests again.

**Pull request merged**